common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -77,57 +92,142 @@ func (s *localStoreCertProvider) FetchServerCertificate() (*tls.Certificate, err
 	return s.serverCert, nil
 }
 
+func (s *localStoreCertProvider) fetchServerCertificateFromInline() (*tls.Certificate, error) {
+	if s.tlsSettings.Server.CertData == "" {
+		return nil, nil
+	}
+
+	certBytes, err := base64.StdEncoding.DecodeString(s.tlsSettings.Server.CertData)
+	if err != nil {
+		return nil, fmt.Errorf("TLS public certificate could not be decoded: %w", err)
+	}
+
+	keyBytes, err := base64.StdEncoding.DecodeString(s.tlsSettings.Server.KeyData)
+	if err != nil {
+		return nil, fmt.Errorf("TLS certificate private key could not be decoded: %w", err)
+	}
+
+	serverCert, err := tls.X509KeyPair(certBytes, keyBytes)
+	if err != nil {
+		return nil, fmt.Errorf("loading server tls certificate failed: %v", err)
+	}
+
+	s.serverCert = &serverCert
+	return s.serverCert, nil
+}
+
 func (s *localStoreCertProvider) FetchClientCAs() (*x509.CertPool, error) {
-	if s.tlsSettings.Server.ClientCAFiles == nil {
+	if !s.tlsSettings.Server.InlineData && len(s.tlsSettings.Server.ClientCAFiles) == 0 {
+		return nil, nil
+	}
+	if s.tlsSettings.Server.InlineData && len(s.tlsSettings.Server.ClientCaData) == 0 {
 		return nil, nil
 	}
 
-	// Check under a read lock first
 	s.RLock()
 	if s.clientCAs != nil {
 		defer s.RUnlock()
 		return s.clientCAs, nil
 	}
-	// Not found, manually unlock read lock and move to write lock
+
 	s.RUnlock()
 	s.Lock()
 	defer s.Unlock()
 
+	if s.clientCAs != nil {
+		return s.clientCAs, nil
+	}
+
+	if s.tlsSettings.Server.InlineData {
+		return s.fetchClientCAsFromInline()
+	}
+
+	return s.fetchClientCAsFromFiles()
+}
+
+func (s *localStoreCertProvider) fetchClientCAsFromFiles() (*x509.CertPool, error) {
```

Comment:
Here too I would put the shared logic inside `FetchClientCAs()` and pass explicit arguments to  `fetchClientCAsFromFiles()` and `fetchClientCAsFromInline()`, so that don't even have access to `s`.

Reply:
removed helper function